quantum/split_common/transport.c

```diff
@@ -3,187 +3,124 @@
 #include "matrix.h"
 #include "quantum.h"
 
-#define ROWS_PER_HAND (MATRIX_ROWS/2)
+#define ROWS_PER_HAND (MATRIX_ROWS / 2)
 
 #ifdef RGBLIGHT_ENABLE
-#   include "rgblight.h"
+#  include "rgblight.h"
 #endif
 
 #ifdef BACKLIGHT_ENABLE
-# include "backlight.h"
-  extern backlight_config_t backlight_config;
+#  include "backlight.h"
+extern backlight_config_t backlight_config;
 #endif
 
 #if defined(USE_I2C) || defined(EH)
 
-#include "i2c.h"
+#  include "i2c_master.h"
+#  include "i2c_slave.h"
 
-#ifndef SLAVE_I2C_ADDRESS
-#  define SLAVE_I2C_ADDRESS           0x32
-#endif
+#  define I2C_BACKLIT_START 0x00
+// Need 4 bytes for RGB (32 bit)
+#  define I2C_RGB_START 0x01
+#  define I2C_KEYMAP_START 0x05
 
-#if (MATRIX_COLS > 8)
-#  error "Currently only supports 8 COLS"
-#endif
+#  define TIMEOUT 100
+
+#  ifndef SLAVE_I2C_ADDRESS
+#    define SLAVE_I2C_ADDRESS 0x32
+#  endif
 
 // Get rows from other half over i2c
 bool transport_master(matrix_row_t matrix[]) {
-  int err = 0;
+  i2c_readReg(SLAVE_I2C_ADDRESS, I2C_KEYMAP_START, (void *)matrix, ROWS_PER_HAND * sizeof(matrix_row_t), TIMEOUT);
 
   // write backlight info
-#ifdef BACKLIGHT_ENABLE
-  if (BACKLIT_DIRTY) {
-    err = i2c_master_start(SLAVE_I2C_ADDRESS + I2C_WRITE);
-    if (err) { goto i2c_error; }
-
-    // Backlight location
-    err = i2c_master_write(I2C_BACKLIT_START);
-    if (err) { goto i2c_error; }
-
-    // Write backlight
-    i2c_master_write(get_backlight_level());
-
-    BACKLIT_DIRTY = false;
-  }
-#endif
-
-  err = i2c_master_start(SLAVE_I2C_ADDRESS + I2C_WRITE);
-  if (err) { goto i2c_error; }
-
-  // start of matrix stored at I2C_KEYMAP_START
-  err = i2c_master_write(I2C_KEYMAP_START);
-  if (err) { goto i2c_error; }
-
-  // Start read
-  err = i2c_master_start(SLAVE_I2C_ADDRESS + I2C_READ);
-  if (err) { goto i2c_error; }
-
-  if (!err) {
-    int i;
-    for (i = 0; i < ROWS_PER_HAND-1; ++i) {
-      matrix[i] = i2c_master_read(I2C_ACK);
-    }
-    matrix[i] = i2c_master_read(I2C_NACK);
-    i2c_master_stop();
-  } else {
-i2c_error: // the cable is disconnceted, or something else went wrong
-    i2c_reset_state();
-    return false;
+#  ifdef BACKLIGHT_ENABLE
+  static uint8_t prev_level = ~0;
+  uint8_t        level      = get_backlight_level();
+  if (level != prev_level) {
+    i2c_writeReg(SLAVE_I2C_ADDRESS, I2C_BACKLIT_START, (void *)&level, sizeof(level), TIMEOUT);
+    prev_level = level;
   }
-
-#ifdef RGBLIGHT_ENABLE
-  if (RGB_DIRTY) {
-    err = i2c_master_start(SLAVE_I2C_ADDRESS + I2C_WRITE);
-    if (err) { goto i2c_error; }
-
-    // RGB Location
-    err = i2c_master_write(I2C_RGB_START);
-    if (err) { goto i2c_error; }
-
-    uint32_t dword = eeconfig_read_rgblight();
-
-    // Write RGB
-    err = i2c_master_write_data(&dword, 4);
-    if (err) { goto i2c_error; }
-
-    RGB_DIRTY = false;
-    i2c_master_stop();
+#  endif
+
+#  ifdef RGBLIGHT_ENABLE
+  static uint32_t prev_rgb = ~0;
+  uint32_t        rgb      = eeconfig_read_rgblight();
```

Comment:
Does this need to be read out of eeprom rather than directly using the `rgblight_config` struct that the master is currently running?

I've recently implemented an inactivity timeout for rgb underglow like this:
https://gist.github.com/Lenbok/603c66903c048f6879ee0860b37f8a7b 
Ideally I'd like to use `rgblight_toggle_noeeprom()` when the idle timer is responsible for toggling things (both to avoid eeprom writes and also as the inactivity based status probably shouldn't be written to eeprom anyway), but found that I could only get the toggle to get sent to the slave side by using the regular/eeprom version of `rgblight_toggle()`.

Reply:
There's nothing requiring eeprom access, in fact I'd rather there be none. If there's a function to get the config directly (I don't want to be reaching into variables directly) then that can go in without issue... it's just not something I'm concerned with right now.